File: java/stack/RemoveKDigits.java
/*
Given string num representing a non-negative integer num, and an integer k, return the smallest possible integer after removing k digits from num.

Example 1:
    Input: num = "1432219", k = 3
    Output: "1219"
    Explanation: Remove the three digits 4, 3, and 2 to form the new number 1219 which is the smallest.

Example 2:
    Input: num = "10200", k = 1
    Output: "200"
    Explanation: Remove the leading 1 and the number is 200. Note that the output must not contain leading zeroes.

Example 3:
    Input: num = "10", k = 2
    Output: "0"
    Explanation: Remove all the digits from the number and it is left with nothing which is 0.
    

Constraints:
    1 <= k <= num.length <= 105
    num consists of only digits.
    num does not have any leading zeros except for the zero itself.
 */

import java.util.Stack;

public class RemoveKDigits {

    public static String removeKDigits(String num, int k) {
        StringBuilder result = new StringBuilder();
         
        // We have to delete all digits
        if (k >= num.length()) {
            return "0";
        }
        // Nothing to delete
        if (k == 0) {
            return num;
        }
        Stack<Character> s = new Stack<Character>();
 
        for (int i = 0; i < num.length(); i++) {
            char c = num.charAt(i);
           
            // Removing all digits in stack that are greater
            // than this digit(since they have higher
            // weightage)
            while (!s.isEmpty() && k > 0 && s.peek() > c) {
                s.pop();
                k--;
            }
            // ignore pushing 0
            if (!s.isEmpty() || c != '0')
                s.push(c);
        }
       
        // If our k isnt 0 yet then we keep popping out the
        // stack until k becomes 0
        while (!s.isEmpty() && k > 0) {
            k--;
            s.pop();
        }
        if (s.isEmpty())
            return "0";
        while (!s.isEmpty()) {
            result.append(s.pop());
        }
        String str = result.reverse().toString();
 
        return str;
    }

    public static void main(String[] args) {
        System.out.println(removeKDigits("1432219", 3));
        System.out.println(removeKDigits("10200", 1));
        System.out.println(removeKDigits("10", 2));
        System.out.println(removeKDigits("432342", 2));
    }
}

File: javascript/stack/RemoveKDigits.js
/*
Given string num representing a non-negative integer num, and an integer k, return the smallest possible integer after removing k digits from num.

Example 1:
    Input: num = "1432219", k = 3
    Output: "1219"
    Explanation: Remove the three digits 4, 3, and 2 to form the new number 1219 which is the smallest.

Example 2:
    Input: num = "10200", k = 1
    Output: "200"
    Explanation: Remove the leading 1 and the number is 200. Note that the output must not contain leading zeroes.

Example 3:
    Input: num = "10", k = 2
    Output: "0"
    Explanation: Remove all the digits from the number and it is left with nothing which is 0.
    

Constraints:
    1 <= k <= num.length <= 105
    num consists of only digits.
    num does not have any leading zeros except for the zero itself.
 */

function removeKDigits(num="", k) {
    if ( k >= num.length) {
        return "0";
    }

    if (k == 0) {
        return num;
    }

    stack = [];

    for(c of num) {
        while (!isEmpty(stack) && k > 0 && peek(stack) > c) {
            stack.pop();
            k--;
        }
        if (!isEmpty(stack) || c != '0') {
            stack.push(c)
        }
    }


    //If after all iteration, k is greater than 0
    while (!isEmpty(stack) && k > 0) {
        k--;
        stack.pop();
    }

    if(isEmpty(stack)){ 
        return "0";
    }

    let result = "";
    stack.reverse();
    while (!isEmpty(stack)) {
        result += stack.pop();
    }

    return result;
}

function peek(stack) {
    if (stack.length > 0)
        return stack[stack.length - 1];
    return 0;
}

function isEmpty(stack) {
    return stack.length <= 0;
}

console.log(removeKDigits("1432219", 3));
console.log(removeKDigits("10200", 1));
console.log(removeKDigits("10", 2));
console.log(removeKDigits("432342", 2));